test/unittests/state_transition.hpp

```diff
@@ -0,0 +1,81 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2023 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+
+#include <evmone/evmone.h>
+#include <gtest/gtest.h>
+#include <test/state/host.hpp>
+
+#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
+
+namespace evmone::test
+{
+using namespace evmone;
+using namespace evmone::state;
+
+/// Fixture to defining test cases in form similar to JSON State Tests.
+///
+/// It takes the "pre" state and produces "post" state by applying the defined "tx" transaction.
+/// Then expectations declared in "except" are checked in the "post" state.
+class state_transition : public testing::Test
+{
+protected:
+    /// The default sender address of the test transaction.
+    /// Private key: 0xa45355879.
+    static constexpr auto Sender = 0xe100DeB58f38F7fd62d14E37e2Fe9ce019Aa001e_address;
+
+    /// The default destination address of the test transaction.
+    static constexpr auto To = 0xc0de_address;
+
+    static constexpr auto Coinbase = 0xc014bace_address;
+
+    static inline evmc::VM vm{evmc_create_evmone()};
+
+    struct ExpectedAccount
+    {
+        bool exists = true;
+        std::optional<uint64_t> nonce;
+        std::optional<intx::uint256> balance;
+        std::unordered_map<bytes32, bytes32> storage;
+    };
+
+    struct Expectation
+    {
+        evmc_status_code status = EVMC_SUCCESS;
+        std::optional<int64_t> gas_used;
+
+        std::unordered_map<address, ExpectedAccount> post;
+    };
+
+
+    BlockInfo block{
+        .gas_limit = 1'000'000,
+        .coinbase = Coinbase,
+        .base_fee = 999,
+    };
+    Transaction tx{
+        .gas_limit = block.gas_limit,
+        .max_gas_price = block.base_fee + 1,
+        .max_priority_gas_price = 1,
+        .sender = Sender,
+        .to = To,
+    };
+    State pre;
+    Expectation expect;
+
+    state_transition()
+    {
+        pre.insert(
```

Comment:
Maybe this initialization should be in `SetUp()` for symmetry with `TearDown()`

Reply:
Moved.